src/main/java/formflow/library/ScreenController.java

```diff
@@ -361,10 +363,16 @@ ModelAndView updateOrCreateIteration(
     return new ModelAndView(viewString);
   }
 
-  private void setIterationIsComplete(String flow, String iterationUuid, FormSubmission formSubmission, Submission submission,
+  private void updateIterationIsCompleteMarker(String flow, String iterationUuid, Submission submission,
       ScreenNavigationConfiguration currentScreen) {
-    Boolean iterationIsComplete = !isNextScreenInSubflow(flow, submission, currentScreen, iterationUuid);
-    formSubmission.getFormData().put("iterationIsComplete", iterationIsComplete);
+
+    boolean isIterationComplete = !isNextScreenInSubflow(flow, submission, currentScreen, iterationUuid);
+
+    Boolean currentStatus = submission.getIterationIsCompleteStatus(currentScreen.getSubflow(), iterationUuid);
+    // if it is already true, leave it alone, otherwise set it
+    if (currentStatus == null || !currentStatus) {
+      submission.setIterationIsCompleteStatus(currentScreen.getSubflow(), iterationUuid, isIterationComplete);
+    }
   }
```

Comment:
I feel strongly that this logic of "only setting this flag if it's unset or false" stays in this class (versus being part of the setter/getting in Submission object). I don't think it should be in the Submission class. That class is concerned with a Submission, not with the business logic around a Submission (though we may have blurred the line elsewhere).